pkg/cloud/services/ec2/instances.go

```diff
@@ -961,6 +982,23 @@ func (s *Service) checkRootVolume(rootVolume *infrav1.Volume, imageID string) (*
 	return rootDeviceName, nil
 }
 
+// ModifyInstanceMetadataOptions modifies the metadata options of the given EC2 instance.
+func (s *Service) ModifyInstanceMetadataOptions(instanceID string, options *infrav1.InstanceMetadataOptions) error {
+	input := &ec2.ModifyInstanceMetadataOptionsInput{
+		HttpEndpoint:            aws.String(string(options.HTTPEndpoint)),
```

Comment:
When creating a cluster, I feel I meet a regression error with latest code related to the options here.  Paste the trace for your awareness...

```
E0308 07:31:06.070432      17 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 499 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x26b8c20, 0x440c890})
	/Users/yikew/Projects/pkg/mod/k8s.io/apimachinery@v0.25.5/pkg/util/runtime/runtime.go:75 +0xe0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:110 +0xf0
panic({0x26b8c20, 0x440c890})
	/usr/local/go/src/runtime/panic.go:890 +0x264
sigs.k8s.io/cluster-api-provider-aws/v2/pkg/cloud/services/ec2.(*Service).ModifyInstanceMetadataOptions(0x400134be60, {0x4001053fc8, 0x13}, 0x0)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/pkg/cloud/services/ec2/instances.go:971 +0x34
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).ensureInstanceMetadataOptions(0x40005e06c0, {0x2ffb9b0, 0x400134be60}, 0x4000f734a0, 0x40011222c0)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:1121 +0xf0
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).reconcileOperationalState(0x40005e06c0, {0x2ffb9b0, 0x400134be60}, 0x40015731a0, 0x4000f734a0)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:616 +0x28c
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).reconcileNormal(0x40005e06c0, {0x40005b5090, 0x40010988e8}, 0x40015731a0, {0x2ffb7f0, 0x400020c310}, {0x3003088, 0x400020c310}, {0x3001be0, 0x400020c310}, ...)
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:588 +0x1624
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).Reconcile(0x40005e06c0, {0x2fee338, 0x4000b39680}, {{{0x40016b7a20, 0x7}, {0x4001030840, 0x16}}})
	/Users/yikew/Projects/src/github.com/kubernetes-sigs/cluster-api-provider-aws/controllers/awsmachine_controller.go:224 +0x93c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x4000330c80, {0x2fee338, 0x4000b39680}, {{{0x40016b7a20, 0x7}, {0x4001030840, 0x16}}})
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121 +0xec
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x4000330c80, {0x2fee338, 0x4000b39680}, {0x27acba0, 0x40000f5f20})
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320 +0x2f8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x4000330c80, {0x2fee290, 0x40001a4300})
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273 +0x2b8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234 +0xa8
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/Users/yikew/Projects/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:230 +0x400
E0308 07:31:06.070538      17 controller.go:326] "Reconciler error" err="panic: runtime error: invalid memory address or nil pointer dereference [recovered]" controller="awsmachine" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSMachine" AWSMachine="default/c6-control-plane-bswb4" namespace="default" name="c6-control-plane-bswb4" reconcileID=18c05f9f-f837-4366-ab61-e6ce9c225db8
```

Reply:
Weird that this is not caught in E2E tests. @muraee could you please take a look.

Reply:
It won't fail the cluster creation, only raise exceptions in the capa logs.

Reply:
I think when instancemetadata options are not set by user, then it will throw the exceptions, I will raise another PR to make it optional

Reply:
One more thing is we noticed another [issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4117) with awsmachine webhooks so that could be the reason, thius defaults are not getting set.
I will wait for that issue to fix first.

Reply:
thx @Ankitasw, yes it seems the defaults are not getting set for some reason, Although I tried that before and it was working.

Reply:
I believe I hit this exact issue in a separate change. Assuming this is the case, then using the defaulting webhook is not enough. This is the code I think is missing: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4112/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011eR192

`SetDefaults_AWSMachineSpec` needs to be invoked from the AWSMachine controller reconciliation loop. 

The patch site is also very important, see the comment.
```
	// The defaulting must happen before `NewMachineScope` is called since otherwise we keep detecting
	// differences that result in patch operations.

	... do defaulting
```
Specifically, what is suspect is happening is existing AWSMachines that were created before your change can't receive the defaults, since they already exist. So we need to also set the defaults in the controller, at least until the resource is upgraded to another major apiVersion, then we can stop defaulting in the controller.

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4147 should fix it